Objective-C: Add ChannelCredentials to gRPC Call

pquerna opened the pull request:
New properties on `GRPCHost` (`pemRootCerts`, `pemPrivateKey`, `pemCertChain`) allow the channel credentials to be constructed with support for TLS Client certificates.  `GRPCHost` uses these properties when creating a `GRPCChannel`.

The `pathToCertificates` property has been removed, but support for loading the root certificates from the bundle has been updated to support the new interface.

Two notes:
1) I've not been able to run the test cases yet.  I couldn't get the `./tools/run_tests/run_tests.py -l objc` working (see also #5591 )
2) From looking at the other language bindings, it's not clear if the `forHost` pattern is common (i'm most familiar with the grpc-go code)..  Is this pattern wanted for the ObjectiveC bindings?  I tried to follow how OAuth / User-Agent setting was done so far.

grpc-jenkins commented:
Can one of the admins verify this patch?

grpc-jenkins commented:
Can one of the admins verify this patch?

grpc-jenkins commented:
Can one of the admins verify this patch?

grpc-jenkins commented:
Can one of the admins verify this patch?

pquerna commented:
@jcanizales: Any feedback on this PR or the approach?  Happy to go a different direction, but the `forHost` is the only pattern used right now in the Objective C code setting authentication or connection level settings.

jcanizales commented:
Sorry for the delay, Paul! Lately I've been almost 100% focused on fixing https://github.com/grpc/grpc/issues/4064. I'll review your PR today for sure.

jcanizales commented:
The overall design looks good to me. I'm going to go ahead and add inline comments about specifics.

jcanizales commented:
Also - let me know if you're able to run the tests with Cocoapods 0.39. If you rather not downgrade your installed version, I'm happy to fetch the branch and run the tests on my machine.

jcanizales commented on src/objective-c/GRPCClient/GRPCCall+ChannelCredentials.h:
```diff
@@ -0,0 +1,43 @@
+
++ (void)setSslPemRootCerts:(NSString *)pemRootCerts forHost:(NSString *)host;
++ (void)setSslPemPrivateKey:(NSString *)pemPrivateKey forHost:(NSString *)host;
++ (void)setSslPemCertChain:(NSString *)pemCertChain forHost:(NSString *)host;
```
Would someone ever want to set a certificates chain without setting a private key, and vice versa?

pquerna replied:
A client certificate and the associated private key are always used together. For example, in the grpc-go API, they are always used together as a constructor for a credentials object:
https://godoc.org/google.golang.org/grpc/credentials#NewServerTLSFromCert

I'll change it to be one function, that takes both the client certificate and private key at the same time.

jcanizales commented:
Actually, let me ask you a general question if you don't mind. So far, the only use case I've heard of for wanting to set a private key on an iOS app was for a demo app that wouldn't be distributed. The moment an app with an embedded private key is distributed, the key stops being a secret and the purpose is defeated. So what is the use case you have in mind here?

jcanizales commented on src/objective-c/GRPCClient/private/GRPCHost.m:
```diff
@@ -127,8 +181,10 @@ - (GRPCChannel *)channel {
         args[@GRPC_SSL_TARGET_NAME_OVERRIDE_ARG] = _hostNameOverride;
       }
 
+      grpc_channel_credentials *creds = [self channelCredentials:_pemRootCerts privateKey:_pemPrivateKey certChain: _pemCertChain];
```
`grpc_ssl_credentials_create` copies the content of the root certs, certificate chain, and private key. Because the majority of channels will use the default root certs, we want to have only a copy of the corresponding `grpc_channel_credentials` in memory (instead of one per channel).

For the non-default case it's not bad, because we have just one channel per host. But `-[GRPCHost channel]` is called lazily, only when the first RPC to that host is to be sent. It's preferable if we create the `grpc_channel_credentials` object eagerly instead: If there's something wrong with the certificates or the key, debugging is much easier if the failure shows up when the user calls `[GRPCCall set... forHost:...]`, rather than later on upon starting an RPC.

jcanizales commented:
There are issues of style conformance here and there, but let's leave that for the end.

pquerna commented:
@jcanizales Use case:

- For setting a Client Certificate aka **Mutual TLS Authentication**: I'm working on a Mac App. We do not ship a client certificate & private key with the application distribution.  We have two use cases in our Mac app:
  - **gRPC to Helper**: We have a Go based helper that contains portable business logic -- we are wrapping it in native UIs for each OS, eg a native Mac App (ObjectiveC/Swift), a native Windows app (C#).  In this case, on startup, the Helper app gives our Mac App a CA, certificate, and private key it has generated, and operates an HTTPS service on 127.0.0.1 and the Mac App calls to this service.  This is our use case today.
 - **gRPC to Remote Service**:  [The product I work on](https://www.scaleft.com/products/access/) is a Certificate Authority which issues ephemeral short lived client certiciates based on multiple factors for use in authenticating to remote services.  I believe in general there are many ways a user can get a client certificate & private key that is not distributed with an application, and having a way to configure this brings the Objective C API to parity with the other gRPC languages: http://www.grpc.io/docs/guides/auth.html#ssl-tls

- For `setSslPemRootCerts`: Common use case of setting your own trusted Certificate Authorities.  Various reasons that a CA is either not in the Mozilla/CAB list, or it is an internal corporate CA.  This use case exist on iOS or Mac IMO, and is safe/common to bundle with your shipping application.

pquerna commented:
style conformance: Is there a `clang-format` params I should use?   The `./tools/distrib/clang_format_code.sh` script doesn't include `.m` in its file types to format?

jcanizales commented:
Thanks for describing your use cases; it's very useful to me. To make sure I understood it correctly:
1. Mac app asks Go helper "please produce a key and cert for this end user",
2. Go helper creates it and sends it back to the Mac app,
3. The Mac app uses it (or a personal key+cert the user already had) to communicate with your remote server.
In no case the intention of the certificate is to try to identify the client app, rather the end user, correct?

jcanizales commented:
For style conformance we don't have an automated tool yet I'm afraid :disappointed:  We mostly follow [Google's ObjC style guide](https://google.github.io/styleguide/objcguide.xml). @makdharma and I were talking the other day about adding at least a linter (and in an ideal world an auto-formatter). I've just filed https://github.com/grpc/grpc/issues/5703 to track.

pquerna commented:
> In no case the intention of the certificate is to try to identify the client app, rather the end user, correct?

Correct, only using private keys to identify the user or session, not a whole application.

pquerna commented:
@jcanizales Updated patch to orient around a reused `channelCreds` property on `GRPCHost`.  Also changed the external interface to have just two functions: 

- `[GRPCCall setSSLCredentials: ...]` for setting both a client certificate and private key.
- `[GRPCCall setSSLPEMRootCerts:...]` for the use case of just overriding the root CAs.

Which I believe better reflect the common use cases and calling patterns (either setting everything, or just setting the trusted CA List)

jcanizales commented on src/objective-c/GRPCClient/GRPCCall+ChannelCredentials.h:
```diff
@@ -0,0 +1,56 @@
+
+#import "GRPCCall.h"
+
+/** Helpers for setting SSL Trusted Roots, Client Certificates, and Private Key */
```
General question re naming: Should we call it "SSL" or "TLS"? IIUC, "TLS" is more accurate, but more people know right away what "SSL" means.

@jtattermusch: Do you know what other wrapping languages use? C uses SSL. @jboeuf thoughts?

jtattermusch replied:
I think the right terminology is TLS, but C core happens to use SSL in the
names. I think gRPC Java might be using more accurate terminology.

On Fri, Mar 11, 2016 at 7:35 PM, Jorge Canizales <notifications@github.com>
wrote:

> In src/objective-c/GRPCClient/GRPCCall+ChannelCredentials.h
> <https://github.com/grpc/grpc/pull/5592#discussion_r55915234>:
>
> > + * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
> > + * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
> > + * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
> > + * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
> > + * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
> > + * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
> > + * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
> > + * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
> > + * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
> > + * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
> > + *
> > + */
> > +
> > +#import "GRPCCall.h"
> > +
> > +/** Helpers for setting SSL Trusted Roots, Client Certificates, and Private Key */
>
> General question re naming: Should we call it "SSL" or "TLS"? IIUC, "TLS"
> is more accurate, but more people know right away what "SSL" means.
>
> @jtattermusch <https://github.com/jtattermusch>: Do you know what other
> wrapping languages use? C uses SSL. @jboeuf <https://github.com/jboeuf>
> thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/5592/files#r55915234>.
>

jcanizales commented on src/objective-c/GRPCClient/GRPCCall+ChannelCredentials.m:
```diff
@@ -0,0 +1,66 @@
+            withCertChain:(nullable NSString *)pemCertChain
+                  forHost:(nonnull NSString *)host
+                    error:(NSError **)errorPtr {
+  if (!host) {
```
Doesn't `nonnull` save us?

jcanizales commented on src/objective-c/GRPCClient/private/GRPCHost.m:
```diff
@@ -107,6 +115,76 @@ - (grpc_call *)unmanagedCallWithPath:(NSString *)path completionQueue:(GRPCCompl
+    // issuer). Load them as UTF8 and produce an ASCII equivalent.
+    NSString *contentInUTF8 = [NSString stringWithContentsOfFile:path
+                                                        encoding:NSUTF8StringEncoding
+                                                           error:&error];
```
`error:&kDefaultRootsError];` instead?

jcanizales replied:
Actually, now that `GRPCHost` isn't getting a path anymore, but the contents of the file instead, do we really want the `errorPtr`? We're only using it for failing to load the default root certs, and that's not something we should be able to recover from.

pquerna replied:
So, my thought was that having the `errorPtr` is a good API, and might be helpful in the future.

Currently `grpc_ssl_credentials_create` doesn't validate its parameters, it just copies the bytes in.  But, it seems reasonable to believe in a future where there is a function that can validate the root certs / client certs / private key being correct, and therefore being able to add an error-return in a reasonable way.

jcanizales replied:
That makes sense. But it also leads to someone adding their error-checking code, verifying that `*errorPtr` stays `nil`, and still getting a parse error later on and cursing us because we told him everything was fine :) So there's pros and cons. Defaulting to the simpler interface seems cautious then; we can always add a version with `errorPtr` later, when the C layer helps us.

jcanizales commented on src/objective-c/GRPCClient/private/GRPCHost.m:
```diff
@@ -127,9 +205,14 @@ - (GRPCChannel *)channel {
+        _channel = [GRPCChannel secureChannelWithHost:_address
+                                          credentials:_channelCreds
+                                          channelArgs:args];
+      }
```
It's preferable to leave calls to other classes out of the `synchronized` block, so that a reader can easily tell there's no possibility of deadlock. Something like:
```objc
grpc_channel_credentials *channelCreds = NULL;
@synchronized(self) {
  if (!_channelCreds) {
    // Use default root certificates.
    [self setSSLCredentials:nil withPrivateKey:nil withCertChain:nil error:nil];
  }
  channelCreds = _channelCreds;
}
_channel = [GRPCChannel secureChannelWithHost:_address
                                  credentials:channelCreds
                                  channelArgs:args];
```

pquerna replied:
Copying the `channelCreds` pointer to a local pointer creates a race condition with the `grpc_channel_credentials_release` in `setSSLCredentials` above.

Options:
1) If we could use `grpc_channel_credentials_ref`, we could safely copy the channel credentials locally, but `grpc_channel_credentials_ref` is not an exported function from the C-core.  The C-core doesn't export direct ref functions currently, so doesn't seem like the right path.
2) Could add reference counting in GRPCHost to guard the calling of `grpc_channel_credentials_release`
3) Document that `setSSLCredentials` is not  multi-threaded thread safe. (and remove the existing `@synchronized` from it
4) keep the `@synchronized` with the current scope.

2 is just a few more lines of code, 4 is the most simple.  Thoughts? Other ideas?

jcanizales replied:
My feeling is that doing (2) by adding a `GRPCChannelCredentials` class that creates the C object and manages its life (and has `[GRPCChannelCredentials defaultCredentials]` and whatnot) is The Right Way (TM). But it might not be worth the hassle if we're just improving readability with it. So I'm happy with (4) too.

jcanizales commented:
Hi Paul, I'm happy with the new design so I went ahead and added inline suggestions for the implementation. It looks great overall! Just a few things to comment.

pquerna commented:
@jcanizales: rebased to master and changed method names to `setTLS` instead of `setSSL`.

PR now includes https://github.com/grpc/grpc/pull/5737 and https://github.com/grpc/grpc/pull/5736 so that I can compile/test it, but submited them as separate PRs as they are both 1-2 liners.

pquerna commented:
fixed & updated rebased so #5737 isn't there.

jtattermusch commented:
this is ok to test

jcanizales commented on src/objective-c/GRPCClient/GRPCCall+ChannelCredentials.h:
```diff
@@ -0,0 +1,56 @@
+           withPrivateKey:(nullable NSString *)pemPrivateKey
+            withCertChain:(nullable NSString *)pemCertChain
+                  forHost:(nonnull NSString *)host
+                    error:(NSError **)errorPtr;
```
@makdharma as I'm on vacation next week, do you want to take over?

My only outstanding concern is whether these methods should have `errorPtr` or not (see the pros and cons elsewhere in the review).
Actually, @jboeuf: Are there plans to, in the future, check the validity of these values when one sets channel credentials? Or is it out of the question for some reason?

After settling that, only a bit of renaming here to make the two methods consistent and follow the style: `s/TLSCrendentials/TLSPEMRootCerts`, and if I remember the style guide correctly, capitalize "PEM" in the parameter names.

jboeuf replied:
@jcanizales we could do some more checking so I would leave the API open to this possibility.

jcanizales commented:
Ah, and the copyright year in `GRPCCall+Tests.m` needs to be changed to `2015-2016` for the tool not to complain.

mgadda commented:
I'd love to see this get merged. Anything I can do to help?

jcanizales commented:
Hi Matt. Paul disappeared for a while, so you can help indeed. I guess the easiest way forward is to copy his branch to your repo and sent it as a PR? There we can make these fixes:
* [Copyright fix](https://github.com/grpc/grpc/pull/5592#issuecomment-198777630) for the tests not to complain.
* s/SSL/TLS everywhere (I think there's only a comment with SSL left)
* [Method and parameter rename](https://github.com/grpc/grpc/pull/5592#discussion_r56754239) for consistency and to follow Apple's acronym conventions.

And we should be good to merge it!

pquerna commented:
I'm happy to update the PR.  Just been focused on other parts of our project, since we have just been vendoring gRPC.

mgadda commented:
@pquerna Sure, if you have time to work on it, that's great. It sounds like there's not that much left to do. Or if not, I'd be happy to help out too (probably with less efficiency, but help nonetheless).

jcanizales commented:
@jboeuf what would be the best way to test this? I don't see an easy map to any of the tests in https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md

jboeuf commented:
@yang-g, @ctiller, @jtattermusch 
I think that we should add interop tests for SSL client certs. Who is the right person to talk to about this?

jcanizales commented:
Friendly ping @pquerna :)

pquerna commented:
Apologies; I don't think I'll be able to get to this for 2-3 weeks, being tied up with other stuff;  CC @mgadda

pquerna commented:
When rebasing this to master, I had a linker error from `grpc_global_wakeup_fd`:

```
duplicate symbol _grpc_global_wakeup_fd in:
    Build/Products/Debug/gRPC/libgRPC.a(ev_posix.o)
    /Build/Products/Debug/gRPC/libgRPC.a(ev_poll_and_epoll_posix.o)
ld: 1 duplicate symbol for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

So, simplest fix was to remove the duplicated grpc_global_wakeup_fd from `ev_posix.c` in 1b9c0a2123bc8e2e3c4a6a7b35e22a7b2b17a69f

jcanizales commented:
Weird! Those symbols were added on Feb 22nd and Feb 25th! How could this not be happening before? :?

jcanizales commented:
And Travis tests for your commit 40f11aa passed alright... ❓ ❓

pquerna commented:
@jcanizales While written farther back in time, #5419 wasn't merged until 28 days ago, April 1

jcanizales commented:
Still plenty of time for the C++ and ObjC linkers to fail :-) In any case, @ctiller has seen ODR violations sneak past linkers before. Craig, this is the PR I just told you.

jcanizales commented:
The PR LGTM, so I'm just waiting for the Jenkins tests to finish.

jcanizales commented:
The C tsan failure is unrelated, and already tracked at https://github.com/grpc/grpc/issues/6186, so I'm merging :) :)
